### DISCUSSION TITLE
手把手教你用 Remix 完成 task 8（拿 10U 奖励）

### DISCUSSION THREAD

**Discussion by linghuccc** [openbuildxyz/Web3-Frontend-Bootcamp#1027] [Materials]
task 8 是一个非常简单的任务，只需要把合约部署在 Morph Holesky Testnet 就可以了。

部署教程 Deploy Contracts on Morph（ https://docs.morphl2.io/docs/build-on-morph/code-examples/deploy-contract-on-morph） 有 2 个部分：Deploy with Hardhat 和 Deploy contracts with Foundry。

介于这个课程有很多合约新手，我补充一下这个教程，增加 Deploy contracts with Remix 的部分。

**第一步：设置小狐狸钱包 Configure MetaMask Wallet；**

首先打开 Meta Mask 钱包，点击网络的下拉菜单：
![m1](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/d3f1369e-8b1e-4f24-8ee1-8ef04f09f300)

选择“Add Network”（添加网络）；
![m2](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/176066f4-df52-441c-aeae-3bdece756fc5)

然后在出现的页面选择“Add a network manually”（手动添加网络）；
![m3](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/f87758c9-7795-4a03-8038-5475b9269816)

根据 Morph docs 教程 https://docs.morphl2.io/docs/build-on-morph/build-on-morph/development-setup 
Step 1: Network Configuration 部分给出的信息，添加 Morph Holesky Testnet 和 Holesky Testnet；
![m4](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/273f6af2-549c-4e42-a937-294c190c9ec4)

添加成功之后，在 MetaMask 钱包 network 的下拉菜单里面，应该可以看到 Morph Holesky Testnet 和 Holesky Testnet。
![m5](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/d26586fe-0ec8-45cb-8719-e752e81cd253)

**第二步：申请测试代币；**

教程 https://docs.morphl2.io/docs/build-on-morph/build-on-morph/development-setup 里面的 Step 3: Acquire Ether ，提供了几个测试代币的水龙头，选择适合你的，申请 ETH 测试币；

申请成功之后，你应该可以在自己的钱包里面看到；
![m6](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/9416cb28-b51a-4779-a8c1-fb4290c2e88d)

此时，ETH 测试币在 Holesky Testnet 网络，根据教程 https://docs.morphl2.io/docs/quick-start/bridge ，你可以将 ETH 测试币跨链到 Morph Holesky Testnet 网络，跨链成功之后，你应该可以在自己的钱包里面看到；
![m7](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/4e4608ae-1f38-4562-915e-ae5dacaebfd6)

**第三步：部署合约；**
打开 https://remix.ethereum.org/ ，编辑合约，编译成功之后，选择 Deploy tab，在左边的 Environment 下拉菜单中，选择 “Injected Provider - MetaMask”；
![r1](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/49406ef6-eb3d-43dc-b58f-10e7d5cb34b1)

点击“Deploy”（部署），此时，MetaMask 钱包会跳出，点击“Confirm”（确定）；
![r2](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/3dc77d17-299b-445d-bad4-192dec4052a3)

如果一切顺利，会显示部署完成。点击合约地址旁边的图标，可以复制合约地址；
![r3](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/606db467-144d-4896-89eb-8931fad372b5)

打开 Morph Holesky Testnet 的 Block Explorer，填入合约地址，可以查看刚部署的合约；
![e1](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/3ca3f6fd-609b-4a29-afa9-f480329e2be9)

**第四步：验证合约；**
此时合约尚未开源，点击 Contract tab；
![e2](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/670fbb50-49d1-4c05-bb6a-2e301b26bde7)

再点击 Verify & publish tab；
![e3](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/61760465-650f-4918-a63d-709e2da8932d)
 
从下拉菜单选择合适的 Verification method （范例合约选择第一项，Flattened） 
![e4](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/26d19a3f-e4bd-454b-b219-bb297cbf373a)

然后从下拉菜单选择对应的 Compiler version （范例合约选择第一项，v0.8.26） 
![e5](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/4e503e96-0745-417a-afb0-74d6c843fdc6)

选择 EVM version，并填入合约代码；
![e6](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/cd495377-f6bf-4433-8be6-522a06fe0f6c)

点击 Verify & publish 之后，如果顺利，会在 Contract tab 旁边看到一个绿色的勾，代表合约已开源；
![e7](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/bc927941-c96e-4fa6-8daa-cac45268f8d9)

合约开源之后，每个人只要输入合约地址，都可以在 block 浏览器查看代码。
![e8](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/7d2a8046-5939-43da-b784-364fed9b4fd9)

以上就是 Deploy contracts with Remix 的全部过程。

如有任何疑问，欢迎大家留言，谢谢。

**Comment by linghuccc** [comment]
有 2 个点要**注意**：

1. 如果合约有 import 其他合约，必须把他们写成一个文件 （flatten 就是这个意思）；
2. Optimization enabled 不要打勾；
![e9](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/59200993/cb9154fa-ccff-40b7-99bc-464e6f45345a)

**Comment by zhangolve** [comment]
![image](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/7566337/7572a72d-a2ed-4780-a86f-6f1e1f9d966e)


为什么我用示例的Lock合约去部署，然后填入Lock.sol 的内容，但是结果却是如图所示，不能验证合约呢

**Reply by linghuccc** [reply]
合约地址发一下，我看看

**Reply by zhangolve** [reply]
我用hardhat verify命令行验证了。

**Reply by zhangolve** [reply]
https://explorer-holesky.morphl2.io/address/0x2537294DC6173A112116bB055E2de3794D1577b2?tab=contract 这是未验证的。

**Reply by zhangolve** [reply]
https://explorer-holesky.morphl2.io/address/0xd3FF8aa41B58F88ADdEA8D916A1Cb4e0F0Ae9CB9 这是我用hardhat命令验证的

**Reply by zhangolve** [reply]
两个合约代码是一样的

**Reply by linghuccc** [reply]
根据 morph 的官方回答，这“是因为部分代码之前已经在我们浏览器上验证过了，所以这次只验证了部分”。
以上是他们的回答，我也无法提供更多解释，不过这个差别不影响这个任务的 bounty。
